### PR TITLE
Integrate seal-oem into COS-Customizer

### PR DESCRIPTION
**Pull request by hegao12**
Bazel now will build a container image (about 33M)
with `veritysetup` stored in builtin_build_ctx.

A subcommand `seal-oem` is added to configure
dm-verity of the OEM partition. When the subcommand
is run, a script name will be written to the state-file
that executes the script in builtin_build_ctx to run
the binary of seal-oem.

After sealing the OEM partition, the system will
not mount "/dev/sda8" after boot. Instead, "/dev/dm-1"
should be mounted at "/usr/share/oem/" manually.

If the subcommand is run, OEM parititon size will be
doubled to store the hash tree. If not, the OEM
partition size will be the same as user input.

Debian snapshot version is also updated to support
the dependencies of `veritysetup`.

The current method to save config.Build to config file
is to use `Save()`, but it doesn't clear the old file
before writing to it. So if the new JSON string of
config.Build is shorter, the tail of the old string will
still be in the config file, corrupting the JSON structure.
`func Update(configFile *os.File, buildConfig *Build) error`
is added to config/config.go. This method first clear the
config file and then use `Save()` to write to it.

### CODE REVIEW

File: cmd/finish_image_build.go [hunk 1]
```diff
@@ -107,33 +107,14 @@ func (f *FinishImageBuild) SetFlags(flags *flag.FlagSet) {
...
-		f.oemFSSize4K = oemSizeBytes >> 12
-		// double the oem size.
-		oemSizeBytes <<= 1
-		oemSizeGB, err := partutil.ConvertSizeToGBRoundUp(strconv.FormatUint(oemSizeBytes, 10) + "B")
-		if err != nil {
-			return fmt.Errorf("invalid format of oem-size: %q, error msg:(%v)", f.oemSize, err)
-		}
-		if (uint64)(f.diskSize)-oemSizeGB < imgSize {
-			return fmt.Errorf("'disk-size-gb' must be at least 'oem-size' x 2 + image size (%dGB)", imgSize)
+		if oemSizeBytes < (16 << 20) {
```
Comment (vaibhav-rustagi) [inline cmd/finish_image_build.go]: Isn't the default size of the OEM partition 16MB? If yes, should OEM partition related workflow be run when same size is provided by User?

File: cmd/finish_image_build.go [hunk 2]
```diff
@@ -171,14 +152,75 @@ func (f *FinishImageBuild) loadConfigs(files *fs.Files) (*config.Image, *config.
...
+			return fmt.Errorf("invalid format of oem-size: %q, error msg:(%v)", buildConfig.OEMSize, err)
+		}
+	} else {
+		if buildConfig.OEMSize == "" {
+			// If need to seal OEM partition and the oem-size is not set,
+			// assume the OEM fs size is 16M as it is in a COS image,
+			// and the OEM partition size is doubled to 32M.
+			// Disk size must be at least 11GB.
+			if buildConfig.DiskSize < 11 {
+				return fmt.Errorf("need extra disk spcae to seal the OEM partition, " +
```
Comment (vaibhav-rustagi) [inline cmd/finish_image_build.go]: space

File: cmd/finish_image_build.go [hunk 3]
```diff
@@ -171,14 +152,75 @@ func (f *FinishImageBuild) loadConfigs(files *fs.Files) (*config.Image, *config.
...
+		if err != nil {
+			return fmt.Errorf("invalid format of oem-size: %q, error msg:(%v)", buildConfig.OEMSize, err)
+		}
+	} else {
+		if buildConfig.OEMSize == "" {
+			// If need to seal OEM partition and the oem-size is not set,
+			// assume the OEM fs size is 16M as it is in a COS image,
+			// and the OEM partition size is doubled to 32M.
+			// Disk size must be at least 11GB.
+			if buildConfig.DiskSize < 11 {
```
Comment (vaibhav-rustagi) [inline cmd/finish_image_build.go:188]: Shouldn't the case of setting OEMSeal option to true be dependent on OEMSize? For example: as a customer, if I didn't provide the OEMSize and set the  SealOEM, then as a user I won't be getting the expected stateful partition size.

File: cmd/finish_image_build.go [hunk 4]
```diff
@@ -171,14 +152,75 @@ func (f *FinishImageBuild) loadConfigs(files *fs.Files) (*config.Image, *config.
...
+			return fmt.Errorf("invalid format of oem-size: %q, error msg:(%v)", buildConfig.OEMSize, err)
+		}
+		buildConfig.OEMFSSize4K = oemSizeBytes >> 12
+		// double the oem size.
+		oemSizeBytes <<= 1
+	}
+	// Since we allow user input like "500M", and the "resize-disk" API can only take GB as input,
+	// the oem-size is rounded up to GB to make sure there is enough space.
+	// Extra space will be taken by the stateful partition.
+	oemSizeGB, err := partutil.ConvertSizeToGBRoundUp(strconv.FormatUint(oemSizeBytes, 10) + "B")
```
Comment (vaibhav-rustagi) [inline cmd/finish_image_build.go:212]: Use imgSize

File: data/builtin_build_context/seal_oem.sh [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+
+# get user input from meatadata
```
Comment (vaibhav-rustagi) [inline data/builtin_build_context/seal_oem.sh]: metadata

File: data/builtin_build_context/seal_oem.sh [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+oem_fs_size_4k="$(/usr/share/google/get_metadata_value \
+attributes/OEMFSSize4K)"
+sudo mount -o remount,exec /var
```
Comment (vaibhav-rustagi) [inline data/builtin_build_context/seal_oem.sh:22]: why is this needed? Does the seal_oem.bin runs under /var?

File: tools/partutil/helpers_test.go [hunk 7]
```diff
@@ -161,49 +161,4 @@ func TestConvertSizeToGBRoundUpFails(t *testing.T) {
 	}
 }
 
-func TestConvertSizeToGBRoundUpPasses(t *testing.T) {
```
Comment (vaibhav-rustagi) [inline tools/partutil/helpers_test.go]: Why is this test deleted?

File: tools/seal_oem_partition.go [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+	if err != nil {
+		return fmt.Errorf("cannot read partUUID of /dev/sda8")
+	}
+	if err := appendDMEntryToGRUB(grubPath, devName, partUUID, hash, salt, oemFSSize4K); err != nil {
+		return fmt.Errorf("error in appending entry to grub.cfg, input:oemFSSize4K=%d, "+
+			"error msg:(%v)", oemFSSize4K, err)
+	}
+	log.Println("kernel command line modified.")
+	if err := removeVeritysetupImage(imageID); err != nil {
+		return fmt.Errorf("cannot remove veritysetup container or image, error msg:(%v)", err)
```
Comment (vaibhav-rustagi) [inline tools/seal_oem_partition.go]: Maybe this should not be error and can be a warning as everything was done by the container?

File: tools/seal_oem_partition.go [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+			"id=%q, error msg: (%v)", imageID, err)
+	}
+	return nil
+}
+
+// mountEFIPartition mounts the EFI partition (/dev/sda12)
+// and returns the path where grub.cfg is at.
+func mountEFIPartition() (string, error) {
+	var tmpDirBuf bytes.Buffer
+	cmd := exec.Command("mktemp", "-d")
```
Comment (vaibhav-rustagi) [inline tools/seal_oem_partition.go]: How about creating a directory at /mnt/stateful_partition/ and removing that in the cleanup?

File: data/builtin_build_context/seal_oem.sh (hunk 6 above)
Comment (hegao12) [reply data/builtin_build_context/seal_oem.sh:22]: the working directory is in /var/lib including the builtin_build_context where the binay lives

File: cmd/finish_image_build.go (hunk 4 above)
Comment (hegao12) [reply cmd/finish_image_build.go:212]: this 10 is not imgSize, but the base of converting uint64 to string

File: tools/seal_oem_partition.go (hunk 8 above)
Comment (hegao12) [reply tools/seal_oem_partition.go]: If it fails to delete the image, the test expecting only one ubuntu image will fail

File: cmd/finish_image_build.go (hunk 1 above)
Comment (rkolchmeyer) [inline cmd/finish_image_build.go]: nit: Maybe leave a comment here saying that 16 << 20 is the default size of the OEM partition

File: cmd/finish_image_build.go [hunk 10]
```diff
@@ -171,14 +152,75 @@ func (f *FinishImageBuild) loadConfigs(files *fs.Files) (*config.Image, *config.
...
 	outputImageConfig.Labels = f.labels.m
 	outputImageConfig.Licenses = f.licenses.l
 	outputImageConfig.Family = f.imageFamily
 	return sourceImageConfig, buildConfig, outputImageConfig, nil
 }
 
+func validateOEM(buildConfig *config.Build) error {
+	// The default size of a COS image (imgSize) is assumed to be 10GB.
+	const imgSize uint64 = 10
+	var sizeErrorMsg string = ""
```
Comment (rkolchmeyer) [inline cmd/finish_image_build.go]: nit: the '= ""' isn't necessary. The default value of string variables is "", so you can just do `var sizeErrorMsg string`.

File: cmd/finish_image_build.go [hunk 11]
```diff
@@ -171,14 +152,75 @@ func (f *FinishImageBuild) loadConfigs(files *fs.Files) (*config.Image, *config.
...
+	var oemSizeBytes uint64 = 0
```
Comment (rkolchmeyer) [inline cmd/finish_image_build.go]: nit: Same here, the default value for uint64 is 0, so you can do `var oemSizeBytes uint64`.

File: cmd/finish_image_build.go (hunk 3 above)
Comment (rkolchmeyer) [reply cmd/finish_image_build.go:188]: IMO sealing the OEM partition without extending it is a valid use case. We just need to document that additional space must be specified by the user to account for the hashtree. Since the user needs to provide the total size of the disk, user expectations in regards to stateful partition size should be met.

File: cmd/finish_image_build.go (hunk 2 above)
Comment (rkolchmeyer) [inline cmd/finish_image_build.go]: nit: spcae -> space

File: cmd/finish_image_build.go (hunk 4 above)
Comment (vaibhav-rustagi) [reply cmd/finish_image_build.go:212]: thanks for the clarification.

File: cmd/finish_image_build.go [hunk 12]
```diff
@@ -171,14 +152,75 @@ func (f *FinishImageBuild) loadConfigs(files *fs.Files) (*config.Image, *config.
...
+	oemSizeGB, err := partutil.ConvertSizeToGBRoundUp(strconv.FormatUint(oemSizeBytes, 10) + "B")
+	if err != nil {
+		return fmt.Errorf("invalid format of oem-size: %q, error msg:(%v)", buildConfig.OEMSize, err)
+	}
+	if (uint64)(buildConfig.DiskSize) < imgSize+oemSizeGB {
+		return fmt.Errorf(sizeErrorMsg, imgSize)
+	}
+	// shrink OEM size input (rounded down) by 1M to deal with cases
+	// where disk size is 1M smaller than needed.
+	// For example oem-size=1G disk-size-gb=12. In this case the disk size is not large enough.
```
Comment (rkolchmeyer) [inline cmd/finish_image_build.go]: I think it would be good to comment here that the intent is to take the 1M away from the hashtree half in the sealing case, and from the user data in the non-sealing case. That would help explain why we modify buildConfig.OEMSize after we compute and use oemSizeBytes.

File: config/config.go [hunk 13]
```diff
@@ -68,11 +68,26 @@ type Build struct {
...
 	OEMSize     string
 	OEMFSSize4K uint64
+	SealOEM     bool
 	GPUType     string
 	Timeout     string
 	GCSFiles    []string
 }
 
+// Update clears the config file and then saves the new config.Build.
+func Update(configFile *os.File, buildConfig *Build) error {
```
Comment (rkolchmeyer) [inline config/config.go]: The name "Update" typically implies that the file isn't blindly overwritten, and is instead modified in a somewhat non-destructive way. Also, this function is specific to the build config (and does not work for other configs). For these reasons, I think it might be better to name this "SaveBuildConfigToFile" or "OverwriteBuildConfig".

File: config/config.go [hunk 14]
```diff
@@ -68,11 +68,26 @@ type Build struct {
...
+	if _, err := configFile.Seek(0, 0); err != nil {
+		return fmt.Errorf("cannot seek config file, error msg:(%v)", err)
```
Comment (rkolchmeyer) [inline config/config.go]: nit: Since this function is specific to the build config, it might be good to say "build config" instead of "config" in the error message.

File: data/builtin_build_context/seal_oem.sh [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+
+# get user input from meatadata
+oem_fs_size_4k="$(/usr/share/google/get_metadata_value \
+attributes/OEMFSSize4K)"
```
Comment (rkolchmeyer) [inline data/builtin_build_context/seal_oem.sh]: nit: I think this would be a bit more readable if the line continuation were indented (I don't think the indentation should cause problems)

File: data/builtin_build_context/seal_oem.sh [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+sudo mount -o remount,exec /var
+sudo chmod 777 ./seal_oem.bin
+sudo ./seal_oem.bin "${oem_fs_size_4k}"
```
Comment (rkolchmeyer) [inline data/builtin_build_context/seal_oem.sh]: nit: Can you add a newline to the end of the file?

File: testing/seal_oem_test/preload_test.cfg [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+          echo "/mnt/stateful_partition/hello contains ${actual} instead of ${expected}"
+          echo "testHello fail"
+          RESULT="fail"
+          return
+        fi
+        echo "testHello pass"
+      }
+
+      testOEMHello(){
+        if [[ ! -f /mnt/stateful_partition/hello ]]; then
```
Comment (rkolchmeyer) [inline testing/seal_oem_test/preload_test.cfg]: Should this test for /usr/share/oem/hello?

File: tools/partutil/helpers.go [hunk 18]
```diff
@@ -99,3 +102,38 @@ func PartNumIntToString(disk string, partNumInt int) (string, error) {
...
+	// /dev/sda2: PARTLABEL="KERN-A" PARTUUID="de4778dd-c187-8343-b86c-e122f9d234c0"
+	// /dev/sda4: PARTLABEL="KERN-B" PARTUUID="7b8374db-78b2-2748-bab9-a52d0867455b"
+	// /dev/sda5: PARTLABEL="ROOT-B" PARTUUID="8ac60384-1187-9e49-91ce-3abd8da295a7"
+	// /dev/sda11: PARTLABEL="RWFW" PARTUUID="682ef1a5-f7f6-7d42-a407-5d8ad0430fc1"
+	lines := strings.Split(idBuf.String(), "\n")
+	for _, line := range lines {
+		if !strings.HasPrefix(line, partName+":") {
+			continue
+		}
+		for _, content := range strings.Split(line, " ") {
```
Comment (rkolchmeyer) [inline tools/partutil/helpers.go]: Can strings.Fields work here instead of splitting by spaces?

File: tools/seal_oem_partition.go [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+		return fmt.Errorf("cannot umount OEM partition, error msg:(%v)", err)
+	}
+	log.Println("OEM parititon unmounted.")
+	hash, salt, err := veritysetup(imageID, oemFSSize4K)
+	if err != nil {
+		return fmt.Errorf("cannot run veritysetup, input:oemFSSize4K=%d, "+
+			"error msg:(%v)", oemFSSize4K, err)
+	}
+	grubPath, err := mountEFIPartition()
+	log.Println("EFI parititon mounted.")
```
Comment (rkolchmeyer) [inline tools/seal_oem_partition.go]: nit: parititon -> partition

File: tools/seal_oem_partition.go (hunk 8 above)
Comment (rkolchmeyer) [reply tools/seal_oem_partition.go]: Error makes sense here, since we don't want to leak the image onto the user's images/VMs.

File: tools/seal_oem_partition.go (hunk 9 above)
Comment (rkolchmeyer) [inline tools/seal_oem_partition.go]: nit: you can use ioutil.TempDir() intead.

File: tools/seal_oem_partition.go [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+	cmd.Stdout = os.Stdout
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("error in removing docker image, "+
+			"id=%q, error msg: (%v)", imageID, err)
+	}
+	return nil
+}
+
+// mountEFIPartition mounts the EFI partition (/dev/sda12)
+// and returns the path where grub.cfg is at.
```
Comment (rkolchmeyer) [inline tools/seal_oem_partition.go]: Should this say the directory path of grub.cfg?

File: tools/seal_oem_partition.go [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+			"error msg:(%v)", grubPath, grubPath, name, partUUID, oemFSSize4K, hash, salt, err)
+	}
+	lines := strings.Split(string(grubContent), "\n")
+	// add the entry to all kernel command lines containing "dm="
+	for idx, line := range lines {
+		if !strings.Contains(line, "dm=") {
+			continue
+		}
+		startPos := strings.Index(line, "dm=")
+		lineBuf := []rune(line[:len(line)-1])
```
Comment (rkolchmeyer) [inline tools/seal_oem_partition.go:214]: Does this remove a quote or a newline? Can you leave a comment saying which one?

**Review by rkolchmeyer — CHANGES_REQUESTED** [review]

**Comment by rkolchmeyer** [comment]
LGTM, merging.
